packages/stakers/src/stakerComponent.ts

```diff
@@ -0,0 +1,356 @@
+          };
+        } catch (error) {
+          return {
+            status: "error",
+            dnpName,
+            error,
+          };
+        }
+      })
+    );
+  }
+
+  protected async persistSelectedIfInstalled(
+    dnpName: string,
+    dockerNetSvcAliasesMap: Record<string, { regex: RegExp; alias: string }[]>,
+    userSettings?: UserSettingsAllDnps
+  ): Promise<void> {
+    console.log("persisting: ", dnpName);
+    await this.setStakerPkgConfig(
+      dnpName,
+      dockerNetSvcAliasesMap,
+      userSettings
+    );
+  }
+
+  protected async setNew({
+    newStakerDnpName,
+    dockerNetworkName,
+    compatibleClients,
+    dockerNetSvcAliasesMap,
+    userSettings,
+    prevClient,
+  }: {
+    newStakerDnpName: string | null | undefined;
+    dockerNetworkName: string;
+    compatibleClients:
+      | {
+          dnpName: string;
+          minVersion: string;
+        }[]
+      | null;
+    dockerNetSvcAliasesMap: Record<string, { regex: RegExp; alias: string }[]>;
+    userSettings?: UserSettingsAllDnps;
+    prevClient?: string | null;
+  }): Promise<void> {
+    if (!prevClient && !newStakerDnpName) {
+      logs.info("no client selected, skipping");
+      return;
+    }
+
+    const currentPkg = await listPackageNoThrow({
+      dnpName: prevClient || "",
+    });
+
+    if (currentPkg) {
+      if (prevClient && compatibleClients)
+        this.ensureSetRequirements(
+          prevClient,
+          compatibleClients,
+          currentPkg.version
+        );
+      if (prevClient !== newStakerDnpName)
+        await this.unsetStakerPkgConfig(currentPkg, dockerNetworkName);
+    }
+
+    if (!newStakerDnpName) return;
+    // set staker config
+    await this.setStakerPkgConfig(
+      newStakerDnpName,
+      dockerNetSvcAliasesMap,
+      userSettings
+    );
+  }
+
+  /**
+   * Set the staker pkg:
+   * - ensures the staker pkg is installed
+   * - connects the staker pkg to the staker network
+   * - adds the staker network to the docker-compose file
+   * - starts the staker pkg
+   */
+  private async setStakerPkgConfig(
+    dnpName: string,
+    dockerNetSvcAliasesMap: Record<string, { regex: RegExp; alias: string }[]>,
+    userSettings?: UserSettingsAllDnps
+  ): Promise<void> {
+    // ensure pkg installed
+    if (
+      !(await listPackageNoThrow({
+        dnpName,
+      }))
+    )
+      await packageInstall(this.dappnodeInstaller, {
+        name: dnpName,
+        userSettings,
+      });
+
+    const pkg = await listPackage({
+      dnpName,
+    });
+
+    // add staker network to the compose file
+    this.addStakerNetworkToCompose(pkg.dnpName, dockerNetSvcAliasesMap);
+
+    // start all containers
+    //if (!pkg.containers.some((c) => c.running))
+    await dockerComposeUpPackage({ dnpName: pkg.dnpName }, true);
+  }
+
+  /**
+   * Adds the staker network and its fullnode alias to the docker-compose file
+   */
+  private addStakerNetworkToCompose(
```

Comment:
What do you think about something like:

```  /**
   * Adds the staker network and its fullnode alias to the docker-compose file
   */
  private addStakerNetworkToCompose(
    dnpName: string,
    netConfigsToAdd: { [serviceName: string]: ComposeServiceNetworksObj }
  ): void {
    const composeEditor = new ComposeFileEditor(dnpName, false);
    const services = composeEditor.compose.services;
    const rootNetworks = composeEditor.compose.networks || {};

    for (const [serviceName, networkConfig] of Object.entries(netConfigsToAdd)) {

      // Find the service that includes serviceName in its name
      const service = services[serviceName] ||
        Object.entries(services).find(([name]) => name.includes(serviceName))?.[1];

      if (!service) {
        logs.warn(`Service ${serviceName} not found in ${dnpName}, skipping`);
        continue; // TODO: Throw error?
      }

      const serviceNetworks = services[serviceName].networks || {};

      // Merge the network config with the existing service networks
      service.networks = {
        ...serviceNetworks,
        ...networkConfig
      };

      // Ensure all networks are added to the root level
      for (const [networkName] of Object.entries(networkConfig)) {
        if (!rootNetworks[networkName]) {
          rootNetworks[networkName] = {
            external: true
          };
        }
      }
    }

    composeEditor.compose.networks = rootNetworks;

    composeEditor.write();
  }

```

Reply:
Note that here instead of using a regex I just try to find the exact service name and, if not found, I look for the service that includes that name

Reply:
70 lines vs 40 lines

Reply:
committed suggestion see https://github.com/dappnode/DNP_DAPPMANAGER/pull/1927/commits/a2d6ebace974acf48763a0672f53737ff12d4a33